docs/proposals/005_carbon_loading_mechanism.md

```diff
@@ -0,0 +1,252 @@
+           +--------------------------+
+                      |
+                      v
+     +---------------------------------------+
+     | Shell Application                     |
+     | - Bundled Extensions                  |
+     | - References Manifest for Dynamic     |
+     |   Loading at Runtime                  |
+     +---------------------------------------+
+                      |
+                      v
+        Dynamic Loading of Extensions at Runtime
+
+```
+
+### Key Concepts
+
+- Co-located extensions are bundled together with the shell application.
+- Extensions are loaded dynamically at runtime.
+- A manifest file is used to define the extensions that can be loaded.
+- Internal extensions are built first and bundled together with the shell application.
+- External extensions are retrieved, built, tested and bundled with the shell application.
+- Extensions can be hosted on CDNs.
+- Extensions can be loaded in `iFrames`.
+
+### Pros and Cons
+
+- **Pros**:
+
+  - **All Dependencies Resolved at Build Time**: ensures that all extensions, whether internal or external, are fully resolved during the build process. This eliminates CORS issues and prevents individual extension outages.
+  - **Flexibility in Extension Locations**: supports local paths for extensions (e.g., `extensions/xyz/index.js`) for tightly integrated functionalities. Allows external extensions to be hosted on CDNs, enabling scalable and efficient delivery.
+  - **Error Management**:
+    - **Internal Extensions**: Errors can be addressed promptly by committing fixes rather than rolling back, as the main branch remains unaffected.
+    - **External Extensions**: A specific version of an external extension is stored in the `extensions` directory. Errors can be caught during the build or resolved with a re-build. In the worst case, problematic extensions can be removed without affecting the rest of the application.
+  - **Ease of Configuration**: a manifest file serves as the single source of truth for defining accessible extensions. Updates can be made to the manifest without modifying core application logic.
+  - **Support for iFrames**: embedding extensions via iFrames allows for isolated execution of third-party or sandboxed extensions.
+  - **Streamlined Build Process**: the manifest is generated during the build process, ensuring consistency in available extensions and simplifying integration.
+
+- **Cons**:
+  - **Dependency on Manifest File**: the system heavily relies on the accuracy of the manifest file. Errors or misconfigurations could disrupt extension loading.
+  - **Build-Time Configuration**: although dynamic loading reduces deployment frequency, changes to the build process or manifest generation logic still require developer involvement.
+  - **Limited Runtime Flexibility**: since dependencies are resolved at build time, adding new extensions dynamically without rebuilding the application is not feasible.
```

Comment:
Rolling an extension back to its previous version independent of the other extensions will not be possible.

Reply:
good point adding it